compiler/rustc_const_eval/src/interpret/step.rs

```diff
@@ -251,9 +252,14 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
 
             Len(place) => {
                 let src = self.eval_place(place)?;
-                let mplace = self.force_allocation(&src)?;
-                let len = mplace.len(self)?;
-                self.write_scalar(Scalar::from_machine_usize(len, self), &dest)?;
+                if let &ty::Array(_, len) = src.layout.ty.kind() {
```

Comment:
This doesn't look right. It is certainly not right without extensive comments explaining what is happening and why the old code is wrong.

Reply:
It's a fast path that does the right thing (codegen does this opt after all or mir opts do it for codegen).

I wonder if this has any effect on miri at all (the force alloc will just happen later?). If there is no benefit to miri, we should just teach const prop about the Len Rvalue

Reply:
Miri shouldn't usually have fast paths though.

However, `OpTy` also has a `len` method these days, so I think a patch could be made that avoids `force_allocation`. That sounds like a good idea. I don't know if it helps ConstProp though.

Reply:
> Miri shouldn't usually have fast paths though. We want to catch all the UB, not take any shortcuts. ;)

Ok, that makes sense.

> However, `OpTy` also has a `len` method these days, so I think a patch could be made that avoids `force_allocation`. That sounds like a good idea. I don't know if it helps ConstProp though.

I did some testing, and using `OpTy` doesn't seem to help ConstProp.

> If there is no benefit to miri, we should just teach const prop about the Len Rvalue.

There's two ConstProps: `const_prop` and `const_prop_lint` (since #94934). I guess this should be added to both?

And should I force-push this PR, or open a new one? Changing this one might make the existing discussion a bit confusing for future readers.

(sorry for the delay)

Reply:
The issue you want to fix is about the lint, so const_prop_lint is the most relevant one.

> I did some testing, and using OpTy doesn't seem to help ConstProp.

Ah, place_to_op will still want to read the local and that is where ConstProp bails.